hack/cloudfront/readme.md

```diff
@@ -0,0 +1,293 @@
+3. The client receives the redirect and sends a GET to CloudFront. CloudFront verifies the presigned URL.
+4. Our serverless function is invoked by CloudFront and detects that the client IP is NOT from AWS and allows the request to pass through to CloudFront.
+5. CloudFront streams the content back to the client (CloudFront reads transparently from S3 when necessary and this read is free).
+Cost: CloudFront GET request and serverless function execution time + Discounted CloudFront transfer cost.
+
+# 100 Foot View
+We can do even better! What if we had a replica of the app.ci registry in every major region (us-east-1, us-east-2, us-west-1, us-west-2)? 
+If we did, then CloudFront could redirect the client to the "free" app.ci S3 replica bucket within its own region. 
+The problem with this is that when a client pushes content registry.ci.openshift.org, it is only written to one S3 bucket (the one is us-east-1). 
+To replicate the data, [S3 is set up with a replication job](https://aws.amazon.com/getting-started/hands-on/replicate-data-using-amazon-s3-replication/)
+that will copy content to the other regions within 15 minutes.
+
+Therefore, the content pushed to us-east-1 is not guaranteed to be on us-west-2 when a client from us-west-2 makes a request.
+To account for this replication delay, the serverless function will query for the file in the client's regional replica. If it finds the file,
+it will redirect the request to read from the replica S3 bucket directly. If it does not find the file, it will fall back to 
+the next best option. For a client in us-east-2, the next best option is to read content from app.ci's us-east-1 bucket
+(because east-1 to east-2 transfer is less expensive than a CloudFront transfer per GB if there is not NAT gateway
+-- which is true for our build farms). For a client from us-west-2, the next best option is to stream the content from CloudFront.
+
+# 10 Foot View
+Each AWS build farm has its own CloudFront distribution in front of its S3 registry bucket. Our serverless function is configured
+to execute as a behavior for these CloudFront distributions as well, but it behaves slightly differently than when it is running on an app.ci (registry.ci.openshift.org)
+request. The build farm registries are not replicated in every region. So, when a request comes in, the serverless function
+will first check to see if the client is in the us-east-1 region (this is where all build farms reside). If the client is 
+from us-east-1, then the client is redirected to read from the build farm registry's S3 bucket in us-east-1.
+If the client is not in us-east-1, the serverless function will check to see if the app.ci registry S3 replica 
+in the client's region has the requested blob. If it does, the client will be redirected
+to read from the app.ci replica in its region. If the file does not exist, then the serverless function will allow
```

Comment:
adding 
```
(because it was built directly on the build farm -- e.g. as part of a presubmit test -- or the 
15 minute replication has not happened for the image yet)
```